haystack/document_store/faiss.py

```diff
@@ -29,44 +27,79 @@ def __init__(
                         deployment, Postgres is recommended.
         :param index_buffer_size: When working with large datasets, the ingestion process(FAISS + SQL) can be buffered in
                                   smaller chunks to reduce memory footprint.
-        :param vector_size: the embedding vector size.
-        :param faiss_index: load an existing FAISS Index.
+        :param vector_dim: the embedding vector size.
+        :param faiss_index_factory_str: Create a new FAISS index of the specified type.
+                                        The type is determined from the given string following the conventions
+                                        of the original FAISS index factory.
+                                        Recommended options:
+                                        - "Flat" (default): Best accuracy (= exact). Becomes slow and RAM intense for > 1 Mio docs.
+                                        - "HNSW": Graph-based heuristic. If not further specified,
+                                                  we use a RAM intense, but more accurate config:
+                                                  HNSW256, efConstruction=256 and efSearch=256
+                                        - "IVFx,Flat": Inverted Index. Replace x with the number of centroids aka nlist.
+                                                          Rule of thumb: nlist = 10 * sqrt (num_docs) is a good starting point.
+                                        For more details see:
+                                        - Overview of indices https://github.com/facebookresearch/faiss/wiki/Faiss-indexes
+                                        - Guideline for choosing an index https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index
+                                        - FAISS Index factory https://github.com/facebookresearch/faiss/wiki/The-index-factory
+                                        Benchmarks: XXX
+        :param faiss_index: Pass an existing FAISS Index, i.e. an empty one that you configured manually
+                            or one with docs that you used in Haystack before and want to load again.
         """
-        self.vector_size = vector_size
-        self.faiss_index = faiss_index
+        self.vector_dim = vector_dim
+
+        if faiss_index:
+            self.faiss_index = faiss_index
+        else:
+            self.faiss_index = self._create_new_index(vector_dim=self.vector_dim, index_factory=faiss_index_factory_str, **kwargs)
+
         self.index_buffer_size = index_buffer_size
         super().__init__(url=sql_url)
 
-    def _create_new_index(self, vector_size: int, index_factory: str = "HNSW4"):
-        index = faiss.index_factory(vector_size + 1, index_factory)
+    def _create_new_index(self, vector_dim: int, index_factory: str = "Flat", metric_type=faiss.METRIC_INNER_PRODUCT, **kwargs):
+        if index_factory == "HNSW" and metric_type == faiss.METRIC_INNER_PRODUCT:
```

Comment:
Sorry for adding comment now, but adding here for more context -

   I think better to limit faiss related internals in this class. As user of this class already getting freedom to pass self configured index in class via `faiss_index` and `faiss_index_factory_str` parameters of the constructor. It will keep this class clean.

Another point how prevent `index.add()` function from breaking? With `faiss_index_factory_str="IDMap,Flat"` parameter vector `add` call will fail.  Which was addressed in #385 as follows:

```
total_indices = self.size()
        ids = np.arange(total_indices, total_indices + len(embeddings), 1, dtype=np.int64)

        if isinstance(self.faiss_index, IndexIDMap):
            self.faiss_index.add_with_ids(vectors_to_add, ids)
        else:
            self.faiss_index.add(vectors_to_add)
```

What is your view on this [comment](https://github.com/deepset-ai/haystack/pull/385#issuecomment-697003090)

Reply:
Yes, normally I would agree that we should limit it to FAISS internals. However, HNSW Flat with IP is not giving the same results when initialized via the index factory. The performance is totally off. See https://github.com/facebookresearch/faiss/issues/1434
Therefore, we need a special case handling here anyway and do a direct init without the factory. I then took the chance and added "good" defaults for the other params. I believe most users will be overwhelmed by deciding on a proper index type for their use case. I believe most will use either just "Flat" or our preconfigured "HNSW" variant. We will also benchmark those two and try to tune the default HNSW config to fit the most common use cases.

Reply:
Regarding the IDMap. What would be the use case here? I think the FAISS id handling is rather something internal that is not interesting for most haystack users. Therefore, I don't think we need to support it,  but maybe I am missing a good use case here :thinking: ?

Reply:
> Yes, normally I would agree that we should limit it to FAISS internals. However, HNSW Flat with IP is not giving the same results when initialized via the index factory. The performance is totally off. See facebookresearch/faiss#1434

Yes agree with you. We can move this part to another function to make it more explicit. I am also able to reproduce facebookresearch/faiss#1434 even tried by compiling fresh from top of the branch. I suspect some issue related to default value of `metric_type` in `struct`. They recently added support for IP metric type for `HNSW` so suspect some places they might have missed it.

Reply:
> Regarding the IDMap. What would be the use case here? I think the FAISS id handling is rather something internal that is not interesting for most haystack users. Therefore, I don't think we need to support it, but maybe I am missing a good use case here 🤔 ?

Actually `IDMap` have very good use case it abstract faiss internal id with external id. Hence one can `update` or `delete` embedding from faiss index store. So far what I understood that internally embeddings are sequentially indexed or the corresponding vector_id. Hence if one id is removed in between then it causes update of vector_id for all the embeddings coming after that record. 
So far haystack only allow add method for `faiss document_store` (Elastic search don't have this issue). And if user want to update/remove particular document then they need to delete all documents and then add them again. There another solution to mark document to be removed as stale but it will not prevent faiss search to return stale document vector_id. This is what I covered in this [comment](https://github.com/deepset-ai/haystack/pull/385#issuecomment-697003090).

Reply:
More links related IDMap

Reply:
@tholor Have you got use case of using IDMap? BTW Milvus handle these case better 
https://dzone.com/articles/how-milvus-realizes-the-delete-function

Now they are planning to support multi storage options. https://docs.google.com/document/d/1iwwLH4Jtm3OXIVb7jFYsfmcbOyX6AWZKaNJAaXC7-cw/edit